### DISCUSSION TITLE
Cannot create Context()

### DISCUSSION THREAD

**Discussion by hillaryan** [pyudev/pyudev#502] [Q&A]
Hello!

I feel like I am out of my depth on the interaction between `mamba` and `find_library` as it is used by `pyudev`. Any input would be appreciated:

```
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyudev
>>> pyudev.__version__
'0.24.1'
>>> pyudev.udev_version()
249
>>> pyudev.Context()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hillary/miniforge3/envs/main/lib/python3.11/site-packages/pyudev/core.py", line 59, in __init__
    self._libudev = load_ctypes_library("udev", SIGNATURES, ERROR_CHECKERS)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hillary/miniforge3/envs/main/lib/python3.11/site-packages/pyudev/_ctypeslib/utils.py", line 52, in load_ctypes_library
    lib = CDLL(library_name, use_errno=True)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hillary/miniforge3/envs/main/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: /home/hillary/miniforge3/envs/main/lib/udev: cannot read file data: Is a directory
>>> 
```
Hazarding a guess, it looks to me like pyudev is trying to open `udev`, but `mamba` has pointed it to `/home/hillary/miniforge3/envs/main/lib/udev` which is not the library's location.

Looking into this more: 
```
>>> find_library("udev")
'/home/hillary/miniforge3/envs/main/lib/udev'
```
Whereas
```
ldconfig -p | grep libudev

libudev.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libudev.so.1
libudev.so.1 (libc6) => /lib/i386-linux-gnu/libudev.so.1
libudev.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libudev.so
```

Any suggestions on a way to make pyudev locate the correct library?

**Comment by mulkieran** [comment]
@hillaryan Unfortunately, you'll have to complain to ctypes I think. See docs here: https://docs.python.org/3/library/ctypes.html#finding-shared-libraries . You _might_ want to try to work backward, i.e., find some arg that you can pass to ```find_library``` that finds the library, but it might not work out. You might want to try to find out what ```find_library``` is using to locate the library, by, e.g., using a debugger just on that function.

**Comment by hillaryan** [comment]
Sounds good. For the moment, I have moved `/home/hillary/miniforge3/envs/main/lib/udev` -> `/home/hillary/miniforge3/envs/main/lib/udev_tmp` and everything works. At some point, I will need to do the legwork you suggest.
